File: test/test_unit_oriented_line.rb
require_relative "../lib/rgfa.rb"
require "test/unit"

TestUnit ||= Module.new

class TestUnit::OrientedLine < Test::Unit::TestCase

  @@sym = :a
  @@ref = "S\ta\t*\txx:Z:1.0".to_rgfa_line
  @@invalid_sym = :"a\ta"
  @@invalid_ref = []
  @@ol_s = RGFA::OrientedLine.new(@@sym, :+)
  @@ol_r = RGFA::OrientedLine.new(@@ref, :-)
  @@ol_inv = RGFA::OrientedLine.new(@@ref, :x)
  @@ol_s_str = "a+"
  @@ol_r_str = "a-"

  def test_new
    assert_nothing_raised { RGFA::OrientedLine.new(@@sym, :+) }
    # no validation on creation:
    assert_nothing_raised { RGFA::OrientedLine.new(@@invalid_sym, :X) }
  end

  def test_OL
    assert_equal(@@ol_s, OL[:a, :+])
    assert_kind_of(RGFA::OrientedLine, OL[:a, :+])
    assert_raise(ArgumentError) {OL[:a]}
    assert_raise(ArgumentError) {OL[:a,:+,:+]}
  end

  def test_to_oriented_line
    assert_equal(@@ol_s, @@ol_s.to_oriented_line)
    assert(@@ol_s.eql?(@@ol_s.to_oriented_line))
    assert_equal(@@ol_s, "a+".to_oriented_line)
    assert_kind_of(RGFA::OrientedLine, "a+".to_oriented_line)
    assert_equal(@@ol_s, [:a, :+].to_oriented_line)
    assert_kind_of(RGFA::OrientedLine, [:a, :+].to_oriented_line)
  end

  def test_line
    assert_equal(@@sym, @@ol_s.line)
    assert_equal(@@ref, @@ol_r.line)
    ol2 = RGFA::OrientedLine.new(@@sym, :-)
    ol2.line = @@ref
    assert_equal(@@ref, ol2.line)
  end

  def test_orient
    assert_equal(:+, @@ol_s.orient)
    assert_equal(:-, @@ol_r.orient)
    ol2 = RGFA::OrientedLine.new(@@sym, :+)
    ol2.orient = :-
    assert_equal(:-, ol2.orient)
  end

  def test_name
    assert_equal(@@sym, @@ol_s.name)
    assert_equal(@@sym, @@ol_r.name)
  end

  def test_validate
    assert_nothing_raised { @@ol_s.validate }
    assert_nothing_raised { @@ol_r.validate }
    ol1 = RGFA::OrientedLine.new(:a, :X)
    ol2 = RGFA::OrientedLine.new(@@invalid_ref, :+)
    ol3 = RGFA::OrientedLine.new(@@invalid_sym, :+)
    assert_raise(RGFA::ValueError) { ol1.validate }
    assert_raise(RGFA::TypeError) { ol2.validate }
    assert_raise(RGFA::FormatError) { ol3.validate }
  end

  def test_invert
    inv_s = @@ol_s.invert
    assert_equal(@@ol_s.line, inv_s.line)
    assert_equal(:-, inv_s.orient)
    inv_r = @@ol_r.invert
    assert_equal(@@ol_r.line, inv_r.line)
    assert_equal(:+, inv_r.orient)
    assert_raise(RGFA::ValueError) { @@ol_inv.invert }
  end

  def test_to_s
    assert_equal(@@ol_s_str, @@ol_s.to_s)
    assert_equal(@@ol_r_str, @@ol_r.to_s)
  end

  def test_equal
    ol2 = RGFA::OrientedLine.new(@@sym, :+)
    ol3 = RGFA::OrientedLine.new(@@ref, :-)
    assert(ol2 == @@ol_s)
    assert(ol3 == @@ol_r)
    # only name and orient equivalence is checked, not line
    assert(@@ol_r != @@ol_s)
    assert(@@ol_r.invert == @@ol_s)
    # equivalence to string
    assert(@@ol_s == "a+")
    assert(@@ol_r == "a-")
    # equivalence to symbol
    assert(@@ol_s == :"a+")
    assert(@@ol_r == :"a-")
    # equivalence to array
    assert(@@ol_s == [:a, :+])
    assert(@@ol_r == [:a, :-])
  end

  def test_block
    ol = RGFA::OrientedLine.new(:a, :+)
    assert_nothing_raised {ol.line = :b}
    assert_nothing_raised {ol.orient = :-}
    ol.block
    assert_raise(RGFA::RuntimeError) {ol.line = :b}
    assert_raise(RGFA::RuntimeError) {ol.orient = :-}
    ol.unblock
    assert_nothing_raised {ol.line = :b}
    assert_nothing_raised {ol.orient = :-}
  end

  def test_delegate_methods
    assert_equal("*", @@ol_r.field_to_s(:sequence))
    assert_equal("1.0", @@ol_r.xx)
    ol = RGFA::OrientedLine.new("S\ta\t*".to_rgfa_line, "+")
    ol.set("xx", 1)
    assert_equal("S\ta\t*\txx:i:1", ol.line.to_s)
  end

end

File: tests/test_unit_oriented_line.py
import unittest
import gfapy

class TestUnitOrientedLine(unittest.TestCase):

  def test_init(self):
    a = gfapy.OrientedLine("a","+")
    # no validation on creation: (invalid orientation)
    gfapy.OrientedLine("a","*")
    # no validation on creation: (invalid line name)
    gfapy.OrientedLine("a\ta","+")
    b = gfapy.OrientedLine("a+")
    self.assertEqual(a, b)
    c = gfapy.OrientedLine(["a","+"])
    self.assertEqual(a, c)
    self.assertRaises(IndexError, gfapy.OrientedLine, [])
    self.assertRaises(IndexError, gfapy.OrientedLine, ["a"])
    # nothing raised, if too many args are provided (further are ignored)
    gfapy.OrientedLine(["a", "+", 1])

  def test_properties(self):
    a = gfapy.OrientedLine("a", "+")
    self.assertEqual("a", a.line)
    self.assertEqual("+", a.orient)
    self.assertEqual("a", a.name)
    s = gfapy.Line("S\tb\t*\txx:Z:1.0")
    a.line = s
    self.assertEqual(s, a.line)
    self.assertEqual("b", a.name)
    self.assertEqual("+", a.orient)
    a.orient = "-"
    self.assertEqual(s, a.line)
    self.assertEqual("-", a.orient)

  def test_validate(self):
    gfapy.OrientedLine("a","+").validate()
    gfapy.OrientedLine(gfapy.Line("S\tb\t*\txx:Z:1.0"),
                       "-").validate()
    self.assertRaises(gfapy.ValueError,
        gfapy.OrientedLine("a","*").validate)
    self.assertRaises(gfapy.TypeError,
        gfapy.OrientedLine([],"+").validate)
    self.assertRaises(gfapy.FormatError,
        gfapy.OrientedLine("a\ta","+").validate)


  def test_inverted(self):
    os = gfapy.OrientedLine("a", "+")
    inv_os = os.inverted()
    self.assertEqual("a", inv_os.line)
    self.assertEqual("+", os.orient)
    self.assertEqual("-", inv_os.orient)
    s = gfapy.Line("S\tb\t*\txx:Z:1.0")
    os = gfapy.OrientedLine(s, "-")
    inv_os = os.inverted()
    self.assertEqual(s, inv_os.line)
    self.assertEqual("-", os.orient)
    self.assertEqual("+", inv_os.orient)
    os = gfapy.OrientedLine("a", "*")
    self.assertRaises(gfapy.ValueError, os.invert)

  def test_str(self):
    self.assertEqual("a-", str(gfapy.OrientedLine("a","-")))
    s = gfapy.Line("S\tb\t*\txx:Z:1.0")
    self.assertEqual("b+", str(gfapy.OrientedLine(s,"+")))

  def test_equal(self):
    a = gfapy.OrientedLine("a", "+")
    b = gfapy.OrientedLine(gfapy.Line("S\ta\t*"), "+")
    c = gfapy.OrientedLine("a", "-")
    self.assertEqual(a, b)
    self.assertNotEqual(a, c)
    # line itself is not checked for equiv, only name:
    b2 = gfapy.OrientedLine(gfapy.Line("S\ta\tCACAC"), "+")
    self.assertEqual(b, b2)
    # equivalence to string:
    self.assertEqual("a+", a)
    self.assertEqual("a+", b)
    self.assertEqual(a, "a+")
    self.assertEqual(b, "a+")
    # equivalence to list:
    self.assertEqual(a, ["a", "+"])
    self.assertEqual(b, ["a", "+"])
    self.assertEqual(["a", "+"], a)
    self.assertEqual(["a", "+"], b)

  def test_block(self):
    a = gfapy.OrientedLine("a", "+")
    a._block()
    with self.assertRaises(gfapy.RuntimeError):
      a.line = "b"
    a._unblock()
    a.line = "b"

  def test_delegate_methods(self):
    ol = gfapy.OrientedLine(gfapy.Line("S\ta\tCACAC"), "+")
    self.assertEqual("CACAC", ol.sequence)
    self.assertEqual("CACAC", ol.field_to_s("sequence"))
    ol.set("xx", 1)
    self.assertEqual("xx:i:1", ol.field_to_s("xx", True))
